File: backend-node/src/controllers/AuthController.ts
import { Request, Response } from "express";
import { AppDataSource } from "../data-source";
import { User, UserRole, UserStatus } from "../entities/User";
import { PasswordResetToken } from "../entities/PasswordResetToken";
import * as bcrypt from "bcrypt";
import * as jwt from "jsonwebtoken";
import { v4 as uuidv4 } from "uuid";
import * as dotenv from "dotenv";

dotenv.config();

const JWT_SECRET = process.env.JWT_SECRET || "secret";
const JWT_EXPIRATION_MS = parseInt(process.env.JWT_EXPIRATION_MS || "86400000");

export class AuthController {
    static async signin(req: Request, res: Response) {
        const { username, password } = req.body;
        const userRepository = AppDataSource.getRepository(User);

        const user = await userRepository.findOne({ where: { username } });

        if (!user || !(await bcrypt.compare(password, user.password))) {
            return res.status(401).json({ message: "Error: Invalid username or password!" });
        }

        const token = jwt.sign(
            { sub: user.username, role: user.role },
            JWT_SECRET,
            { expiresIn: JWT_EXPIRATION_MS / 1000 }
        );

        return res.json({
            token,
            id: Number(user.id),
            username: user.username,
            email: user.email,
            role: user.role,
            fullName: user.fullName,
            phoneNumber: user.phoneNumber,
            address: user.address,
            registrationDate: user.registrationDate
        });
    }

    static async signup(req: Request, res: Response) {
        const { username, email, password, role, fullName, phoneNumber, address } = req.body;
        const userRepository = AppDataSource.getRepository(User);

        if (await userRepository.findOne({ where: { username } })) {
            return res.status(400).json({ message: "Error: Username is already taken!" });
        }

        if (await userRepository.findOne({ where: { email } })) {
            return res.status(400).json({ message: "Error: Email is already in use!" });
        }

        const hashedPassword = await bcrypt.hash(password, 10);
        const user = userRepository.create({
            username,
            email,
            password: hashedPassword,
            role: (role as UserRole) || UserRole.ROLE_MEMBER,
            fullName,
            phoneNumber,
            address,
            status: UserStatus.ACTIVE,
            registrationDate: new Date()
        });

        await userRepository.save(user);

        return res.json({ message: "User registered successfully!" });
    }

    static async forgotPassword(req: Request, res: Response) {
        const { email } = req.body;
        const userRepository = AppDataSource.getRepository(User);
        const tokenRepository = AppDataSource.getRepository(PasswordResetToken);

        const user = await userRepository.findOne({ where: { email } });

        if (!user) {
            return res.json({ message: "If the email exists, a password reset link has been sent." });
        }

        await tokenRepository.delete({ user: { id: user.id } });

        const token = uuidv4();
        const resetToken = tokenRepository.create({
            token,
            user,
            expiryDate: new Date(Date.now() + 3600000), // 1 hour
            used: false
        });

        await tokenRepository.save(resetToken);

        console.log(`[Development] Password reset token for ${email}: ${token}`);

        return res.json({ message: "If the email exists, a password reset link has been sent." });
    }

    static async resetPassword(req: Request, res: Response) {
        const { token, newPassword } = req.body;
        const tokenRepository = AppDataSource.getRepository(PasswordResetToken);
        const userRepository = AppDataSource.getRepository(User);

        const resetToken = await tokenRepository.findOne({
            where: { token },
            relations: ["user"]
        });

        if (!resetToken || resetToken.used || resetToken.expiryDate < new Date()) {
            return res.status(400).json({ message: "Invalid or expired reset token." });
        }

        const user = resetToken.user;
        user.password = await bcrypt.hash(newPassword, 10);
        await userRepository.save(user);

        resetToken.used = true;
        await tokenRepository.save(resetToken);

        return res.json({ message: "Password has been reset successfully!" });
    }

    static async googleLogin(req: Request, res: Response) {
        // Placeholder implementation matching Spring Boot
        const email = "google.user@example.com";
        const username = `google_${uuidv4().substring(0, 8)}`;

        const userRepository = AppDataSource.getRepository(User);
        let user = await userRepository.findOne({ where: { email } });

        if (!user) {
            user = userRepository.create({
                username,
                email,
                password: await bcrypt.hash(uuidv4(), 10),
                role: UserRole.ROLE_MEMBER,
                status: UserStatus.ACTIVE,
                registrationDate: new Date()
            });
            await userRepository.save(user);
        }

        const token = jwt.sign(
            { sub: user.username, role: user.role },
            JWT_SECRET,
            { expiresIn: JWT_EXPIRATION_MS / 1000 }
        );

        return res.json({
            token,
            id: Number(user.id),
            username: user.username,
            email: user.email,
            role: user.role,
            fullName: user.fullName,
            phoneNumber: user.phoneNumber,
            address: user.address,
            registrationDate: user.registrationDate
        });
    }
}

File: backend/src/main/java/com/library/library_management/controller/AuthController.java
package com.library.library_management.controller;

import com.library.library_management.dto.AuthDto.*;
import com.library.library_management.model.PasswordResetToken;
import com.library.library_management.model.User;
import com.library.library_management.repository.PasswordResetTokenRepository;
import com.library.library_management.repository.UserRepository;
import com.library.library_management.security.JwtUtils;
import com.library.library_management.service.EmailService;
import jakarta.transaction.Transactional;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.Optional;
import java.util.UUID;

@CrossOrigin(origins = "*", maxAge = 3600)
@RestController
@RequestMapping("/api/auth")
public class AuthController {
    @Autowired
    AuthenticationManager authenticationManager;

    @Autowired
    UserRepository userRepository;

    @Autowired
    PasswordEncoder encoder;

    @Autowired
    JwtUtils jwtUtils;

    @Autowired
    EmailService emailService;

    @Autowired
    PasswordResetTokenRepository passwordResetTokenRepository;

    @PostMapping("/signin")
    public ResponseEntity<?> authenticateUser(@Valid @RequestBody LoginRequest loginRequest) {

        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(loginRequest.getUsername(), loginRequest.getPassword()));

        SecurityContextHolder.getContext().setAuthentication(authentication);
        String jwt = jwtUtils.generateJwtToken(authentication);

        org.springframework.security.core.userdetails.User userDetails = (org.springframework.security.core.userdetails.User) authentication
                .getPrincipal();
        User user = userRepository.findByUsername(userDetails.getUsername()).get();

        return ResponseEntity.ok(new JwtResponse(jwt,
                user.getId(),
                user.getUsername(),
                user.getEmail(),
                user.getRole().name(),
                user.getFullName(),
                user.getPhoneNumber(),
                user.getAddress(),
                user.getRegistrationDate()));
    }

    @PostMapping("/signup")
    public ResponseEntity<?> registerUser(@Valid @RequestBody SignupRequest signUpRequest) {
        if (userRepository.existsByUsername(signUpRequest.getUsername())) {
            return ResponseEntity.badRequest().body(new MessageResponse("Error: Username is already taken!"));
        }

        if (userRepository.existsByEmail(signUpRequest.getEmail())) {
            return ResponseEntity.badRequest().body(new MessageResponse("Error: Email is already in use!"));
        }

        User user = User.builder()
                .username(signUpRequest.getUsername())
                .email(signUpRequest.getEmail())
                .password(encoder.encode(signUpRequest.getPassword()))
                .fullName(signUpRequest.getFullName())
                .phoneNumber(signUpRequest.getPhoneNumber())
                .address(signUpRequest.getAddress())
                .status(User.UserStatus.ACTIVE)
                .registrationDate(java.time.LocalDate.now())
                .build();

        String strRole = signUpRequest.getRole();
        if (strRole == null) {
            user.setRole(User.Role.ROLE_MEMBER);
        } else {
            user.setRole(User.Role.valueOf(strRole));
        }

        userRepository.save(user);

        return ResponseEntity.ok(new MessageResponse("User registered successfully!"));
    }

    @PostMapping("/forgot-password")
    @Transactional
    public ResponseEntity<?> forgotPassword(@Valid @RequestBody ForgotPasswordRequest request) {
        Optional<User> userOpt = userRepository.findByEmail(request.getEmail());

        if (userOpt.isEmpty()) {
            // Return success even if email doesn't exist (security best practice)
            return ResponseEntity.ok(new MessageResponse("If the email exists, a password reset link has been sent."));
        }

        User user = userOpt.get();

        // Delete any existing tokens for this user
        passwordResetTokenRepository.deleteByUserId(user.getId());

        // Generate new token
        String token = UUID.randomUUID().toString();
        PasswordResetToken resetToken = PasswordResetToken.builder()
                .token(token)
                .user(user)
                .expiryDate(LocalDateTime.now().plusHours(1))
                .used(false)
                .build();

        passwordResetTokenRepository.save(resetToken);

        // Send email (logs to console in development)
        emailService.sendPasswordResetEmail(user.getEmail(), token);

        return ResponseEntity.ok(new MessageResponse("If the email exists, a password reset link has been sent."));
    }

    @PostMapping("/reset-password")
    @Transactional
    public ResponseEntity<?> resetPassword(@Valid @RequestBody ResetPasswordRequest request) {
        Optional<PasswordResetToken> tokenOpt = passwordResetTokenRepository.findByToken(request.getToken());

        if (tokenOpt.isEmpty()) {
            return ResponseEntity.badRequest().body(new MessageResponse("Invalid or expired reset token."));
        }

        PasswordResetToken resetToken = tokenOpt.get();

        if (resetToken.isExpired() || resetToken.isUsed()) {
            return ResponseEntity.badRequest().body(new MessageResponse("Invalid or expired reset token."));
        }

        // Update user password
        User user = resetToken.getUser();
        user.setPassword(encoder.encode(request.getNewPassword()));
        userRepository.save(user);

        // Mark token as used
        resetToken.setUsed(true);
        passwordResetTokenRepository.save(resetToken);

        return ResponseEntity.ok(new MessageResponse("Password has been reset successfully!"));
    }

    @PostMapping("/google")
    public ResponseEntity<?> googleLogin(@Valid @RequestBody GoogleLoginRequest request) {
        // For development: Create/login user with Google ID
        // In production, you would verify the idToken with Google's API

        // Extract email from token (simplified for demo)
        // In production: Use Google's token verification library
        String email = "google.user@example.com"; // Placeholder
        String username = "google_" + UUID.randomUUID().toString().substring(0, 8);

        Optional<User> existingUser = userRepository.findByEmail(email);
        User user;

        if (existingUser.isPresent()) {
            user = existingUser.get();
        } else {
            // Create new user from Google account
            user = User.builder()
                    .username(username)
                    .email(email)
                    .password(encoder.encode(UUID.randomUUID().toString())) // Random password
                    .role(User.Role.ROLE_MEMBER)
                    .status(User.UserStatus.ACTIVE)
                    .build();
            userRepository.save(user);
        }

        // Generate JWT token
        Authentication authentication = new UsernamePasswordAuthenticationToken(
                user.getUsername(), null, null);
        SecurityContextHolder.getContext().setAuthentication(authentication);
        String jwt = jwtUtils.generateJwtToken(authentication);

        return ResponseEntity.ok(new JwtResponse(jwt,
                user.getId(),
                user.getUsername(),
                user.getEmail(),
                user.getRole().name(),
                user.getFullName(),
                user.getPhoneNumber(),
                user.getAddress(),
                user.getRegistrationDate()));
    }
}
